lib/kitchen/driver/vagrant.rb

```diff
@@ -250,6 +246,48 @@ def check_vagrant_version
             " Please upgrade to version #{MIN_VER} or higher from #{WEBSITE}."
         end
       end
+
+      def set_state(state)
+        hash = vagrant_ssh_config
+
+        state[:hostname] = hash["HostName"]
+        state[:username] = config[:username] || hash["User"]
+        state[:password] = config[:password] || 'vagrant'
+        state[:ssh_key] = hash["IdentityFile"]
+        state[:port] = port_from_config
+        if config[:communicator] == "ssh"
+          state[:port] =  hash["Port"]
+        end
+        refresh_forwarded_port(state)
+      end
+
+      def port_from_config
+        port = config[:port]
+        if port.nil?
+          guest_port = config[:guest_port] || instance.transport.default_port
+          if !config[:network].empty?
+            forwards = config[:network].select do |net|
+              net[0] == "forwarded_port" && net[1][:guest] == guest_port
+            end
+            forwards.each do |forward|
+              port = forward[1][:host]
+            end
+          end
+        end
+        port || instance.transport.default_port
+      end
+
+      # Get forwarded_port from Provider
+      #
+      # Working with: VirtualBox
+      def refresh_forwarded_port(state)
```

Comment:
Someone help me with ruby syntax - aren't destructive methods normally ended in a `!`?  Or is that only if it will raise exceptions?